### PR TITLE
clean up logging and make fatal logs be formatted right too

### PR DESCRIPTION
**Pull request by traetox**
cleaning up logging system to enable relays and also adding the internal ingest point

### CODE REVIEW

File: client/ingest.go [hunk 1]
```diff
@@ -26,6 +26,14 @@ func (c *Client) IngestEntries(entries []types.StringTagEntry) error {
 	return c.putStaticURL(JSON_INGEST_URL, entries)
 }
 
+// IngestInternal is used to perform ingest on internal logs for external components
```
Comment (djfritz) [inline client/ingest.go:29]:
```suggestion
// IngestInternal is used to perform ingest on internal logs for external components.
```

File: client/ingest.go [hunk 2]
```diff
@@ -26,6 +26,14 @@ func (c *Client) IngestEntries(entries []types.StringTagEntry) error {
...
+// Things like the searchagent and other drone controllers can use this to get their
+// internal logs into the the gravwell tag without an ingest connection
```
Comment (djfritz) [inline client/ingest.go:31]:
```suggestion
// internal logs into the the gravwell tag without an ingest connection.
```

File: client/ingest.go [hunk 3]
```diff
@@ -26,6 +26,14 @@ func (c *Client) IngestEntries(entries []types.StringTagEntry) error {
...
+// this API requires admin status
```
Comment (djfritz) [inline client/ingest.go:32]:
```suggestion
// This API requires admin status.
```

File: ingest/log/logging.go [hunk 4]
```diff
@@ -166,13 +171,13 @@ func (l *Logger) RawMode() bool {
...
 func (l *Logger) ready() error {
-	if !l.hot || len(l.wtrs) == 0 {
+	if !l.hot || (len(l.wtrs) == 0 && len(l.rls) == 0) {
 		return ErrNotOpen
 	}
 	return nil
 }
 
-// Add a new writer which will get all the log lines as they are handled
+// AddWriter will add a new writer which will get all the log lines as they are handled
```
Comment (djfritz) [inline ingest/log/logging.go:180]:
```suggestion
// AddWriter will add a new writer which will get all the log lines as they are handled.
```

File: ingest/log/logging.go [hunk 5]
```diff
@@ -186,6 +191,20 @@ func (l *Logger) AddWriter(wtr io.WriteCloser) error {
 	return nil
 }
 
+// AddRelay will add a new relay which will get all log entries as they are handled
```
Comment (djfritz) [inline ingest/log/logging.go:194]:
```suggestion
// AddRelay will add a new relay which will get all log entries as they are handled.
```

File: ingest/log/logging.go [hunk 6]
```diff
@@ -186,6 +191,20 @@ func (l *Logger) AddWriter(wtr io.WriteCloser) error {
...
+func (l *Logger) AddRelay(r Relay) error {
+	if r == nil {
+		return errors.New("Invalid relay, is nil")
```
Comment (djfritz) [inline ingest/log/logging.go:197]:
```suggestion
		return errors.New("Nil relay")
```

File: ingest/log/logging.go [hunk 7]
```diff
@@ -204,6 +223,24 @@ func (l *Logger) DeleteWriter(wtr io.Writer) error {
 	return nil
 }
 
+// DeleteRelay removes a relay from the logger
```
Comment (djfritz) [inline ingest/log/logging.go:226]:
```suggestion
// DeleteRelay removes a relay from the logger.
```

File: ingest/log/logging.go [hunk 8]
```diff
@@ -204,6 +223,24 @@ func (l *Logger) DeleteWriter(wtr io.Writer) error {
...
+func (l *Logger) DeleteRelay(rl Relay) error {
+	if rl == nil {
+		return errors.New("Invalid relay, is nil")
```
Comment (djfritz) [inline ingest/log/logging.go:229]:
```suggestion
		return errors.New("Nil relay")
```

**Review by djfritz — CHANGES_REQUESTED** [review]

**Comment by traetox** [comment]
@djfritz PTAL

**Review by djfritz — APPROVED** [review]
LGTM

**Pull request merged**